### PR TITLE
Add initial functionality for Cobbler formula

### PR DESCRIPTION
**Pull request by SchoolGuy**
This PR should include the basic functionality of the Cobbler Formula:

- Install the Cobbler package
- Add the community repository in case Cobbler is not available for your OS (with an explicit flag)
- Template and manage the settings file
- Install optional services like DHCP and DNS in case the settings desire it
- Enable a user to pass an override configuration file that ignores the Salt configuration

Uninstallation Cobbler in a way that no trace is left is not part of this PR. This PR focuses on getting Cobbler running.

### CODE REVIEW

**Comment by SchoolGuy** [comment]
For CentOS 8 we should use https://github.com/saltstack-formulas/epel-formula to enable Cobbler to be used.

**Comment by tacerus** [comment]
I wonder if it might make sense to not template the configuration files which use standard formats, but to rather use the YAML serializer for .yaml configuration failes and [`ini_manage`](https://docs.saltproject.io/en/latest/ref/states/all/salt.states.ini_manage.html) states for .ini ones. Of course, there might be special cases where custom handling is needed, but from the files I notice in this patch, the only "special" part seems to be the comment block used in the file header.

File: cobbler/package/clean.sls [hunk 1]
```diff
@@ -9,6 +9,12 @@
 include:
   - {{ sls_config_clean }}
 
+{% if pkg.communityrepo.enabled %}
+cobbler-package-clean-repository-removed:
+  pkgrepo.absent:
+    - name: cobbler_community_repo
+{% endif %}
```
Comment (tacerus) [inline cobbler/package/clean.sls]: I might be overlooking something, but if `pkg.communityrepo` is _enabled_, should it really be removed? It seems to conflict with it being installed in `repository.sls`.

**Comment by tacerus** [comment]
Minor nitpick, it would be nice to use `{%-` instead of `{%` for Jinja operations in the sls files, as to not print lots of empty lines while running in debug mode.

File: cobbler/package/clean.sls (hunk 1 above)
Comment (SchoolGuy) [reply cobbler/package/clean.sls]: As far as the upstream methodology goes it is desired to have a `<formula>.clean` state that is responsible for uninstallation of the formula. As such I believe it should not be a conflict. The "substate" `clean` should be present in other areas as well.

**Comment by SchoolGuy** [comment]
@opoplawski I am trying to get Cobbler installed on CentOS 8 Stream, AlmaLinux 8 & RockyLinux 8. Is the module that is available for CentOS 8 not available for those distros?

Edit 1: Version 9 of the corresponding OSes are sadly not available in their (Saltstack) images repository.

Edit 2: I found the issue. The module is not needed anymore. Now I need to find out why Salt is refusing to install Cobbler. locally without Salt this just works fine.

**Comment by SchoolGuy** [comment]
Splitout PRs:
- #5 
- #6
- #7
- #8

**Comment by SchoolGuy** [comment]
Force push after rebase on `main` branch.

**Comment by SchoolGuy** [comment]
@tacerus Do you have any good idea how to get the installed version of Cobbler and depending on that choose a configuration file? The configuration schema is enforced by Cobbler and for the RHEL 8-based distros we have an older version of Cobbler that differs in the schema. We need a template for each Cobbler version the formula supports and use it appropriately.

The only idea (which wouldn't work out of the box) is that we define the template in the pillar that has to be used and the users need to know which template they need to use. I'm not too fond of that idea, to be honest though.

Edit: I also asked in the Salt Formula Slack since I didn't see any possible inspiration in the official formulas I looked at. - https://saltstackcommunity.slack.com/archives/C7LG8SV54/p1700730778200969

**Comment by SchoolGuy** [comment]
Executing a command doesn't work as expected since the package is not installed at the time of rendering the template. I need to think more about this...

**Comment by tacerus** [comment]
You could skip the command if the package is not installed, but it's of course not nice having to apply states twice to get the desired result.

I have a far stretched idea: running the command inside the template file used in cobbler-config-file-file-managed, and then `{%- include files_switch(...) %}` inside there. I think the included template file might only be rendered once `cobbler-config-file-file-manage` is executed (which already has a `require` on the package state), unlike the `file.sls`, which is rendered before the states run. Not sure about this though, haven't tried it.

Edit: on a second thought, this cannot work, as it needs to render it in advance to find a possible diff.

Regarding 9341e730b1cf449599893fafa11fbda08e71e47b, it might be something for a future patch - you could write the state in pure Python (`#!py` renderer) allowing you to import and use `configparser` directly, avoiding having to load a subshell with Python from within Python.

**Comment by SchoolGuy** [comment]
@tacerus so in short we have no way to do this in a single run?

Do you think it is viable to ask the user to hardcode the version in the formula? I think not but well I don't have any other idea except running the formula twice...

Running the state twice makes testing in the current CI next to impossible...

**Comment by tacerus** [comment]
It might be possible using an Orchestration state, but I think it would be overkill for this.

I do think setting it using the pillar is viable. What about setting the versions shipped in the distribution packages as defaults?

**Comment by SchoolGuy** [comment]
Okay this is a great idea. I do think we have a way forward then. I'll get right to it after I am in front of my keyboard.

**Comment by SchoolGuy** [comment]
@tacerus I have fixed it partly after a good chunk of waiting (lol 9 months). Do you have an idea why the CI fails for CentOS 8 here? The upstream testsuite is [successful](https://gitlab.com/saltstack-formulas/epel-formula/-/jobs/4800999728) for CentOS 8 EPEL. Of course, before this is ready for merging I need to do some more fine-tuning.

**Comment by SchoolGuy** [comment]
With CentOS 9 EPEL installs successfully but the Cobbler module is not available.

**Comment by SchoolGuy** [comment]
@tacerus I think this PR is now ready for a review from you. It isn't finished but should be good for feedback. After your feedback I will need to verify and fix more OSes then what the CI is running. Lastly I will need to check what templated variables are missing in the different templates.

The CentOS 8 issues are due to CentOS moving its repositories to vault.

**Comment by SchoolGuy** [comment]
All current parameters from `default.yaml` should now be utilized to configure Cobbler. Testing in a real-world infrastructure wasn't done yet and validating how good the formula works on other OSes is also an open point. I believe for the scope of this PR the work is done.

**Comment by tacerus** [comment]
Hi,

I think this is ok for an initial patch. Some things could be improved, like iterating over the pillar keys instead of declaring all the configuration options individually (to simplify the code but also to guard against unset pillar values).

Testing all the supported distributions would sure be good but can be added over time.

**Review by tacerus — APPROVED** [review]

**Pull request merged**